src/suffix.c

```diff
@@ -124,55 +128,211 @@ void deleteSuffixTrie(Trie *trie, const char *str, uint32_t len) {
+      processSuffixData(data, sufCtx);
+    }
+  }
+}
+
+
+/***********************************************************************************
+*                                    Wildcard                                      *
+************************************************************************************/
+
+/* Breaks wildcard at '*'s and find the best token to get iterate the suffix trie  */
+int Suffix_ChooseToken(const char *str, size_t len, size_t *tokenIdx, size_t *tokenLen) {
+  int runner = 0;
+  int i = 0;
+  int init = 0;
+  while (i < len) {
+    if (str[i] != '*') {
+      tokenIdx[runner] = i;
+      init = 1;
+    }
+    while (i < len && str[i] != '*') {
+      ++i;
+    }
+    if (init) {
+      tokenLen[runner] = i - tokenIdx[runner];
+      ++runner;
+    }
+    while (str[i] == '*') {
+      ++i;
+    }
+  }
+
+  // choose best option
+  int score = INT32_MIN;
+  int retidx = -1;
+  for (int i = 0; i < runner; ++i) {
+    if (tokenLen[i] < MIN_SUFFIX) {
+      continue;
+    }
+
+    // 1. long string are likely to have less results
+    // 2. tokens at end of pattern are likely to be more relevant
+    int curScore = tokenLen[i] + i;
+
+    // iterating all children is demanding
+    if (str[tokenIdx[i] + tokenLen[i]] == '*') {
+      curScore -= 5;
+    }
+
+    // this branching is heavy
+    for (int j = tokenIdx[i]; j < tokenIdx[i] + tokenLen[i]; ++j) {
+      if (str[j] == '?') {
+        --curScore;
+      }
+    }
+
+    if (curScore >= score) {
+      score = curScore;
+      retidx = i;
+    }
+  }
+
+  return retidx;
+}
+
+int Suffix_ChooseToken_rune(const rune *str, size_t len, size_t *tokenIdx, size_t *tokenLen) {
+  int runner = 0;
+  int i = 0;
+  int init = 0;
+  while (i < len) {
+    if (str[i] != (rune)'*') {
+      tokenIdx[runner] = i;
+      init = 1;
+    }
+    while (i < len && str[i] != (rune)'*') {
+      ++i;
+    }
+    if (init) {
+      tokenLen[runner] = i - tokenIdx[runner];
+      ++runner;
+    }
+    while (str[i] == (rune)'*') {
+      ++i;
+    }
+  }
+
+  // choose best option
+  int score = INT32_MIN;
+  int retidx = -1;
+  for (int i = 0; i < runner; ++i) {
+    if (tokenLen[i] < MIN_SUFFIX) {
+      continue;
+    }
+
+    // 1. long string are likely to have less results
+    // 2. tokens at end of pattern are likely to be more relevant
+    int curScore = tokenLen[i] + 1;
+
+    // iterating all children is demanding
+    if (str[tokenIdx[i] + tokenLen[i]] == (rune)'*') {
+      curScore -= 5;
+    }
+
+    // this branching is heavy
+    for (int j = tokenIdx[i]; j < tokenIdx[i] + tokenLen[i]; ++j) {
+      if (str[j] == (rune)'?') {
+        --score;
+      }
     }
-  }                              
+
+    if (curScore >= score) {
+      score = curScore;
+      retidx = i;
+    }
+  }
+
+  return retidx;
+}
+
+static int processSuffixData_Wildcard(suffixData *data, SuffixCtx *sufCtx) {
+  if (!data) {
+    return REDISMODULE_OK;
```

Comment:
Can this happened?

Reply:
Yes. If there is no payload.
Merged two functions into one.

Reply:
So it can not happened and if it does then its a bug? I would prefer not to protect it and assert it if this is the case..

Reply:
A node may not be terminal and therefore not have a payload. This is valid.